apps/openmw/mwworld/containerstore.cpp

```diff
@@ -461,27 +461,53 @@ void MWWorld::ContainerStore::addInitialItem (const std::string& id, const std::
 
 void MWWorld::ContainerStore::restock (const ESM::InventoryList& items, const MWWorld::Ptr& ptr, const std::string& owner)
 {
-    // Remove the items already spawned by levelled items that will restock
+    //allowedForReplace - Holds information about how many items from the group were sold;
+    //                    Hence, tells us how many items we need to restock.
+    //allowedForReplace[group] <- How many items we should generate(how many of these were sold)
+    std::map<std::string, int> allowedForReplace;
+
+    //See if any of generated items was sold, take note for further restocking:
     for (std::map<std::string, int>::iterator it = mLevelledItemMap.begin(); it != mLevelledItemMap.end(); ++it)
     {
-        if (count(it->first) >= it->second)
+        int itemCount = count(it->first);
+        //If it was sold
+        if (itemCount < it->second)
+        {
+            //Remove it from the shop
             remove(it->first, it->second, ptr);
+            std::string group = levelledItemGroup(it->first, items);
+            //And add information to the allowedForReplace map
+            std::map<std::string, int>::iterator groupInMap = allowedForReplace.find(group);
+            if( groupInMap == allowedForReplace.end())
+                groupInMap->second = 0;
+            groupInMap->second += std::abs(it->second - itemCount);
+            //Also, remove the record from mLevelledItemMap, if it was the last item in stock.
+            if(!itemCount)
```

Comment:
We also need to adjust mLevelledItemMap if that wasn't the last item in stock. Set the map value to the actual count of items i.e. count(it->first);.